Modules/posixmodule.c

```diff
@@ -8779,6 +8782,14 @@ void
 _Py_closerange(int first, int last)
 {
     first = Py_MAX(first, 0);
+#ifdef HAVE_CLOSE_RANGE
+    if (close_range(first, last, 0) == 0 || errno != ENOSYS) {
```

Comment:
It might be nicer to cache `ENOSYS` the first time we try `close_range` and avoid it in future `_Py_closerange` calls — assuming this is ever used more than once per process, which might not be true.

Reply:
I implemented such cache in the PEP 446 implementation: remind if a syscall supports atomic "O_CLOEXEC" flag. See for example _Py_open_cloexec_works in Python/fileutils.c.

Reply:
In my opinion, I don't think this code path is sensitive enough to really care.

Reply:
ah, nice, I hadn't thought of caching that.  good idea.  (one less failed syscall attempt before the real work likely involving many syscalls gets done in this case; so not a huge deal but still worthwhile)

Reply:
For the subprocess use case, the cache will likely only be first initialized in child processes, each child process has to do the check, and so the cache is useless.

I would mostly benefit to os.closerange().

I'm fine with not adding a cache. I agree that os.closerange() is an uncommon function.